lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,341 @@
+                               .compare(std::get<0>(mem).value()) >= 0))) &&
+                  (!std::get<1>(mem) ||
+                      (std::get<0>(stringRangeValue) &&
+                          ((std::get<0>(stringRangeValue).value())
+                                  .compare(std::get<1>(mem).value()) <= 0)))) ||
+              ((!std::get<0>(mem) ||
+                   (std::get<1>(stringRangeValue) &&
+                       (std::get<1>(stringRangeValue)
+                               .value()
+                               .compare(std::get<0>(mem).value()) >= 0))) &&
+                  (!std::get<1>(mem) ||
+                      (std::get<1>(stringRangeValue) &&
+                          (std::get<1>(stringRangeValue)
+                                  .value()
+                                  .compare(std::get<1>(mem).value()) <= 0)))));
+        });
+
+    if (selectCaseStringRangeValsIter == selectCaseStringRangeVals.end()) {
+      selectCaseStringRangeVals.push_back(stringRangeValue);
+    } else {
+      ReportOverlapCaseValError(
+          std::get<2>(*selectCaseStringRangeValsIter), source);
+    }
+  }
+
+private:
+  using selectCaseStmtTypes =
+      std::vector<std::pair<std::variant<std::int64_t, bool, std::string>,
+          parser::CharBlock>>;
+  selectCaseStmtTypes selectCaseVals;
+  selectCaseStmtTypes::iterator selectCaseValsIter;
+  using selectCaseIntRangeType =
+      std::vector<std::tuple<std::optional<std::int64_t>,
+          std::optional<std::int64_t>, parser::CharBlock>>;
+  selectCaseIntRangeType selectCaseIntRangeVals;
+  selectCaseIntRangeType::iterator selectCaseIntRangeValsIter;
+  using selectCaseStringRangeType =
+      std::vector<std::tuple<std::optional<std::string>,
+          std::optional<std::string>, parser::CharBlock>>;
+  selectCaseStringRangeType selectCaseStringRangeVals;
+  selectCaseStringRangeType::iterator selectCaseStringRangeValsIter;
+  parser::Messages &messages_;
+};
+
+void SelectStmtChecker::Leave(
+    const parser::CaseConstruct &selectCaseConstruct) {
+  const auto &selectCaseStmt{
+      std::get<parser::Statement<parser::SelectCaseStmt>>(
+          selectCaseConstruct.t)};
+  const auto &parsedExpr{
+      std::get<parser::Scalar<parser::Expr>>(selectCaseStmt.statement.t).thing};
+  std::optional<TypeCategory> selectCaseStmtType;
+  int selectCaseStmtKind;
+  bool validCaseStmtType{false};
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (auto type{expr->GetType()}) {
+      selectCaseStmtType = type->category();
+      selectCaseStmtKind = type->kind();
+    }
+  }
+  if (!selectCaseStmtType) {
+    return;
+  }
+  if ((selectCaseStmtType.value() != TypeCategory::Integer) &&
+      (selectCaseStmtType.value() != TypeCategory::Character) &&
+      (selectCaseStmtType.value() != TypeCategory::Logical)) {  // C1145
+    context_.Say(parsedExpr.source,
+        "SELECT CASE expression must be of type CHARACTER, INTERGER, OR LOGICAL"_err_en_US);
```

Comment:
It would be better just to return here. I don't think there is any point in checking the cases when the expression is the wrong type.
Also, "INTERGER" -> "INTEGER".

Reply:
We continue to check for duplicate default case only. We do not consider any other cases.
Here is the condition which checks this: 
242     } else if (validCaseStmtType) {
I didn't notice spell mistake. I will correct them.

Reply:
I still think it would be better just to return. A user that is trying to do select case on some other type needs to rethink what they are doing.